tests/tls_challenge_test.go

```diff
@@ -0,0 +1,258 @@
+	ts.logf("Original value is %q. (Will restore this value on cleanup.)", ts.originalCentralEndpoint)
+
+	ts.setupProxy(proxyNs, ts.originalCentralEndpoint)
+}
+
+func (ts *TLSChallengeSuite) TearDownSuite() {
+	ts.cleanupProxy(ts.cleanupCtx, proxyNs)
+	if ts.originalCentralEndpoint != "" {
+		ts.setDeploymentEnvVal(ts.cleanupCtx, s, sensorDeployment, sensorContainer, centralEndpointVar, ts.originalCentralEndpoint)
+	}
+	// Check sanity after test.
+	waitUntilCentralSensorConnectionIs(ts.T(), ts.cleanupCtx, storage.ClusterHealthStatus_HEALTHY)
+	ts.cancel()
+}
+
+func (ts *TLSChallengeSuite) TestTLSChallenge() {
+	const (
+		proxyServiceName = "nginx-loadbalancer"
+		proxyEndpoint    = proxyServiceName + "." + proxyNs + ":443"
+	)
+
+	ts.logf("Pointing sensor at the proxy...")
+	ts.setDeploymentEnvVal(ts.ctx, s, sensorDeployment, sensorContainer, centralEndpointVar, proxyEndpoint)
+	ts.logf("Sensor will now attempt connecting via the nginx proxy.")
+
+	ts.waitUntilLog(ts.ctx, s, map[string]string{"app": "sensor"}, sensorContainer, "contain info about successful connection",
+		containsLineMatching(regexp.MustCompile("Info: Add central CA cert with CommonName: 'Custom Root'")),
+		containsLineMatching(regexp.MustCompile("Info: Connecting to Central server "+proxyEndpoint)),
+		containsLineMatching(regexp.MustCompile("Info: Established connection to Central.")),
+		containsLineMatching(regexp.MustCompile("Info: Communication with central started.")),
+	)
+	waitUntilCentralSensorConnectionIs(ts.T(), ts.ctx, storage.ClusterHealthStatus_HEALTHY)
+}
+
+func (ts *TLSChallengeSuite) setupProxy(proxyNs string, centralEndpoint string) {
+	name := "nginx-loadbalancer"
+	nginxLabels := map[string]string{"app": "nginx"}
+	nginxTLSSecretName := "nginx-tls-conf"
+	nginxConfigName := "nginx-proxy-conf"
+	ts.logf("Setting up nginx proxy in namespace %q...", proxyNs)
+	ts.createProxyNamespace(proxyNs)
+	ts.installImagePullSecret(proxyNs)
+	ts.createProxyTLSSecret(proxyNs, nginxTLSSecretName)
+	ts.createProxyConfigMap(proxyNs, centralEndpoint, nginxConfigName)
+	ts.createService(ts.ctx, proxyNs, name, nginxLabels, map[int32]int32{443: 8443})
+	ts.createProxyDeployment(proxyNs, name, nginxLabels, nginxConfigName, nginxTLSSecretName)
+	ts.logf("Nginx proxy is now set up in namespace %q.", proxyNs)
+}
+
+func (ts *TLSChallengeSuite) createProxyNamespace(proxyNs string) {
+	_, err := ts.k8s.CoreV1().Namespaces().Create(ts.ctx, &v1.Namespace{ObjectMeta: metaV1.ObjectMeta{Name: proxyNs}}, metaV1.CreateOptions{})
+	if apiErrors.IsAlreadyExists(err) {
+		return
+	}
+	ts.Require().NoError(err, "cannot create proxy namespace %q", proxyNs)
+}
+
+func (ts *TLSChallengeSuite) installImagePullSecret(proxyNs string) {
+	configBytes, err := json.Marshal(config.DockerConfigJSON{
+		Auths: map[string]config.DockerConfigEntry{
+			"https://quay.io": {
+				Username: os.Getenv("REGISTRY_USERNAME"),
```

Comment:
Can this be empty?
https://github.com/stackrox/stackrox/blob/0704c5ec5473b5de50f3f9ffcbea2715fcfcb989/pkg/testutils/centralgrpc/connect_to_central.go#L39

Reply:
Right, it probably could be improved to abort early rather than produce a useless secret. Again, let me follow up on this in next PR.